File: cmd/integration/bookmarks.go
package main

import (
	"log"
	"strings"

	"github.com/playwright-community/playwright-go"
)

// validate the basic logic of the bookmarks page
// execute CRUD operations and check the results
type bookmarksTest struct {
	page playwright.Page
}

func (b *bookmarksTest) validate() {
	prefix := randStr(7)
	b.addEditDeleteBookmark(prefix)
}

func (b *bookmarksTest) addEditDeleteBookmark(p string) {
	err := b.page.Click("#link-add-bookmark")
	assertError("could not click link to add a bookmark; %v", err)

	// create a folder
	b.page.Click("#rd-bookmark-folder")
	b.page.Fill("#txt-bookmark-displaName", p+"__FOLDER")

	err = b.page.Click("#btn-bookmark-save")
	assertError("could not create a new folder; %v", err)

	// wait for the dialog to settle
	b.page.WaitForTimeout(1000)

	// go to the created folder
	item := locateItem(b.page, ".class-bookmark-item", p+"__FOLDER")
	err = item.Click()
	assertError("could not click the link, %v", err)

	b.page.WaitForTimeout(500)

	// ---- create a bookmark ----

	// create a new bookmark item
	err = b.page.Click("#link-add-bookmark")
	assertError("could not click link to add a bookmark; %v", err)

	// create a folder
	b.page.Fill("#txt-bookmark-displaName", p+"__BOOKMARK")
	b.page.Fill("#txt-bookmark-url", "http://www.orf.at")

	err = b.page.Click("#btn-bookmark-save")
	assertError("could not create a new folder; %v", err)

	// wait for the dialog to settle
	b.page.WaitForTimeout(1000)

	// find created bookmark
	locateItem(b.page, ".class-bookmark-item", p+"__BOOKMARK")

	// ---- edit the bookmark ----

	// open the context menu
	locateAndClick(b.page, ".mat-menu-trigger")

	// click the delete button
	locateAndClick(b.page, "#btn-bookmark-edit")

	b.page.Fill("#txt-bookmark-displaName", p+"__BOOKMARK-UPDATE")
	locateAndClick(b.page, "#btn-bookmark-save")

	// settle dialog
	b.page.WaitForTimeout(1000)

	locateItem(b.page, ".class-bookmark-item", p+"__BOOKMARK-UPDATE")

	// ---- delete the bookmark ----

	// open the context menu
	locateAndClick(b.page, ".mat-menu-trigger")

	// click the delete button
	locateAndClick(b.page, "#btn-bookmark-delete")

	// confirm the delete dialog
	locateAndClick(b.page, "#btn-confirm")

	// wait for the dialog to settle
	b.page.WaitForTimeout(1000)

	locateItem(b.page, ".bookmark_item", "no entries available")

	// ---- go outside and search folder ----

	item = locateItem(b.page, ".rootroot", "root")
	item.Click()

	b.page.WaitForTimeout(1000)

	loc, err := b.page.Locator(".bookmark_item")
	assertError("could not locate '.bookmark_item'", err)
	items, _ := loc.ElementHandles()
	for _, item := range items {
		txt, _ := item.InnerText()
		if strings.Contains(txt, p+"__FOLDER") {
			locItem, _ := item.QuerySelector(".mat-menu-trigger")
			locItem.Click()

			// click the delete button
			locateAndClick(b.page, "#btn-bookmark-delete")

			// confirm the delete dialog
			locateAndClick(b.page, "#btn-confirm")
			break
		}
	}

	b.page.WaitForTimeout(1000)

}

func locateAndClick(page playwright.Page, selector string) {
	loc, err := page.Locator(selector)
	assertError("could not get the locator for '"+selector+"'; %v", err)
	loc.Click()
}

func locateItem(page playwright.Page, selector, itemText string) playwright.ElementHandle {
	var (
		itemDisplayName string
		item            playwright.ElementHandle
	)

	loc, err := page.Locator(selector)
	assertError("could not get the locator for class '.class-bookmark-item'; %v", err)
	items, err := loc.ElementHandles()
	assertError("could not get the list of items; %v", err)

	for _, entry := range items {
		name, _ := entry.TextContent()
		log.Printf("found item: %s", name)
		if name == itemText {
			itemDisplayName = name
			item = entry
			break
		}
	}
	assertEqual(itemText, itemDisplayName)

	return item
}

File: testdata/integration/bookmarks.py
import random
import string

from playwright.sync_api import Page, expect


# validate the basic features of the bookmarks logic
def validate(baseURL: str, page: Page):
    prefix = ''.join(random.SystemRandom().choice(string.ascii_letters + string.digits) for _ in range(10))

    # we start by creating a new folder
    page.get_by_test_id('link-add-bookmark').click()
    page.locator('id=rd-bookmark-folder').click()
    page.locator('id=txt-bookmark-displaName').fill(prefix+'__FOLDER')
    page.locator('id=btn-bookmark-save').click()
    page.expect_request('**/api/v1/bookmarks**')

    # find the newly created item
    expect(page.get_by_text(prefix+'__FOLDER')).to_be_visible()
    page.get_by_text(prefix+'__FOLDER').click()

    # create a new bookmark within this folder
    page.get_by_test_id('link-add-bookmark').click()
    page.locator('id=txt-bookmark-displaName').fill(prefix+'__BOOKMARK')
    page.locator('id=txt-bookmark-url').fill('http://www.orf.at')
    page.locator('id=btn-bookmark-save').click()
    page.wait_for_timeout(500)

    # find the newly created item
    expect(page.get_by_text(prefix+'__BOOKMARK')).to_be_visible()
    expect(page.locator('.mat-menu-trigger')).to_be_visible()

    # edit the bookmark
    page.locator('.mat-menu-trigger').click()
    page.locator('id=btn-bookmark-edit').click()
    page.locator('id=txt-bookmark-displaName').fill(prefix+'__BOOKMARK-UPDATE')
    page.locator('id=btn-bookmark-save').click()
    page.wait_for_timeout(500)

    # again find the updated item in the list
    expect(page.get_by_text(prefix+'__BOOKMARK-UPDATE')).to_be_visible()
    expect(page.locator('.mat-menu-trigger')).to_be_visible()

    # delete the created bookmark
    page.locator('.mat-menu-trigger').click()
    page.locator('id=btn-bookmark-delete').click()
    page.locator('id=btn-confirm').click()
    page.wait_for_timeout(500)

    # empty list
    expect(page.get_by_text('no entries available')).to_be_visible()

    # got outside and search for the created folder
    page.locator('.rootroot').click()
    page.wait_for_url(baseURL + '/bookmarks')
    page.wait_for_timeout(500)
    # find the create item
    page.locator('.bookmark_item').filter(has_text=prefix+'__FOLDER').locator('.mat-menu-trigger').click()
    page.locator('id=btn-bookmark-delete').click()
    page.locator('id=btn-confirm').click()
    page.wait_for_timeout(500)
